circuits/src/keccak/mod.rs

```diff
@@ -940,6 +972,197 @@ where
+                &row[i + 1],
+            ));
+        }
+        next_row
+    }
+
+    /// Groups `proof_ids` in groups of 32 bytes (each representing a proof id).
+    /// Keeps the first `num_proof_ids` groups and replaces the rest with zeroes.
+    fn pad_proof_ids(
+        ctx: &mut Context<F>,
+        range: &RangeChip<F>,
+        proof_ids: &[AssignedValue<F>],
+        num_proof_ids: AssignedValue<F>,
+    ) -> Vec<Vec<AssignedValue<F>>> {
+        let mut proof_ids = proof_ids
+            .iter()
+            .chunks(32)
+            .into_iter()
+            .map(|chunk| chunk.into_iter().copied().collect_vec())
+            .collect_vec();
+        let total_num_proof_ids = proof_ids.len();
+        assert_eq!(
+            (total_num_proof_ids & (total_num_proof_ids - 1)),
+            0,
+            "The number of (padded) proof ids must be a power of two in submission id mode"
+        );
+
+        let zero = ctx.load_constant(F::zero());
+        let bitmask = first_i_bits_bitmask(
+            ctx,
+            &range.gate,
+            num_proof_ids,
+            total_num_proof_ids as u64,
+        );
+
+        for (proof_id, bit) in proof_ids.iter_mut().zip(bitmask.iter()) {
+            for byte in proof_id.iter_mut() {
+                *byte = range.gate.select(ctx, *byte, zero, *bit)
+            }
+        }
+
+        proof_ids
+    }
+
+    /// Computes the leaves of the Merkle tree from `proof_ids`,
+    fn compute_leaves(
+        ctx: &mut Context<F>,
+        range: &RangeChip<F>,
+        keccak: &mut KeccakChip<F>,
+        proof_ids: &[AssignedValue<F>],
+        num_proof_ids: AssignedValue<F>,
+    ) -> Vec<Vec<AssignedValue<F>>> {
+        let proof_ids =
+            Self::pad_proof_ids(ctx, range, proof_ids, num_proof_ids);
+        proof_ids
+            .into_iter()
+            .map(|proof_id| Self::compute_leaf(ctx, range, keccak, &proof_id))
+            .collect_vec()
+    }
+
+    /// Computes the submission id from `proof_ids` as bytes.
+    pub(crate) fn compute_submission_id_bytes(
+        ctx: &mut Context<F>,
+        range: &RangeChip<F>,
+        keccak: &mut KeccakChip<F>,
+        proof_ids: &[AssignedValue<F>],
+        num_proof_ids: AssignedValue<F>,
+    ) -> [AssignedValue<F>; 32] {
+        let mut current_row =
+            Self::compute_leaves(ctx, range, keccak, proof_ids, num_proof_ids);
+        let num_leaves = current_row.len();
+        let depth = (num_leaves.ilog2() + 1) as usize;
+        let next_power_of_two = compute_next_power_of_two_bit_decomposition(
+            ctx,
+            range,
+            num_proof_ids,
+            depth,
+        )
+        .into_iter()
+        .map(QuantumCell::from);
+
+        let mut subtree_roots = vec![current_row[0].clone()];
+        while current_row.len() > 1 {
+            current_row = Self::hash_row(ctx, range, keccak, current_row);
+            subtree_roots.push(current_row[0].clone());
+        }
+
```

Comment:
@SupremoUGH Does it make sense to add this:
`assert(subtree_roots.len() == depth)`